File: src/controller.js
'use strict';

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const { send } = require('./ipc');
const { alive, start } = require('./player');
const stations = require('./stations');
const {
  ipcPath,
  stateDir,
  disabledFlag,
  noFocusFlag,
  logFile,
  stationFile,
  volumeFile,
  intentFile,
  activityFile,
  watchdogFile,
  debugEnabled,
  defaultSource,
} = require('./paths');

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const FADE_MS = 180;
const FADE_STEPS = 6;
const ACTIVITY_WINDOW_MS = 8000;
const ACTIVITY_MAX = 6;

function log(message) {
  if (!debugEnabled()) return;
  try {
    fs.mkdirSync(stateDir(), { recursive: true });
    const time = new Date().toISOString().slice(11, 23);
    const ev = process.env.VIBECODE_EVENT || '-';
    fs.appendFileSync(logFile(), `${time} [${ev}] ${message}\n`);
  } catch {
  }
}

async function logAudio(message) {
  if (!debugEnabled()) return;
  const mute = await getProp('mute');
  const paused = await getProp('pause');
  const idle = await getProp('core-idle');
  log(`${message} | audio mute=${mute} pause=${paused} core-idle=${idle}`);
}

function actionToken() {
  return Number(process.env.VIBECODE_TOKEN) || Date.now();
}

function recordIntent(token, action) {
  try {
    if (token < latestToken()) return;
    fs.mkdirSync(stateDir(), { recursive: true });
    fs.writeFileSync(intentFile(), `${token} ${action}`);
  } catch {
  }
}

function latestToken() {
  try {
    return Number(fs.readFileSync(intentFile(), 'utf8').trim().split(' ')[0]) || 0;
  } catch {
    return 0;
  }
}

function superseded(token) {
  return latestToken() > token;
}

async function timed(label, fn) {
  const t0 = Date.now();
  const result = await fn();
  log(`  ${label}: ${Date.now() - t0}ms`);
  return result;
}

function volume() {
  try {
    const saved = Number(fs.readFileSync(volumeFile(), 'utf8').trim());
    if (Number.isFinite(saved)) return saved;
  } catch {
  }
  return Number(process.env.VIBECODE_VOLUME) || 70;
}
const isDisabled = () => fs.existsSync(disabledFlag());

const focusOn = () => !fs.existsSync(noFocusFlag());

const ADAPTIVE_SPREAD = 15;

function adaptiveVolume(base, level) {
  const v = Math.round(base - ADAPTIVE_SPREAD + 2 * ADAPTIVE_SPREAD * level);
  return Math.max(0, Math.min(100, v));
}

function adaptiveEnabled() {
  const v = String(process.env.VIBECODE_ADAPTIVE || '').toLowerCase();
  return !['0', 'false', 'off', 'no'].includes(v);
}

function targetVolume() {
  const base = volume();
  if (!adaptiveEnabled()) return base;
  return adaptiveVolume(base, activityLevel());
}

function source() {
  if (process.env.VIBECODE_SOURCE) return process.env.VIBECODE_SOURCE;
  try {
    const saved = fs.readFileSync(stationFile(), 'utf8').trim();
    if (saved) return saved;
  } catch {
  }
  return defaultSource();
}

async function getProp(name) {
  const reply = await send(ipcPath(), { command: ['get_property', name] });
  if (!reply || reply.error !== 'success') return null;
  return reply.data;
}

async function isHalted() {
  const paused = await getProp('pause');
  if (paused === null) return null;
  if (paused) return true;
  return (await getProp('mute')) === true;
}

async function fade(from, to, guard) {
  for (let i = 1; i <= FADE_STEPS; i += 1) {
    if (guard && superseded(guard)) return false;
    const v = Math.round(from + ((to - from) * i) / FADE_STEPS);
    await send(ipcPath(), { command: ['set_property', 'volume', v] });
    await sleep(FADE_MS / FADE_STEPS);
  }
  return true;
}

const fadeTo = (target) => fade(0, target);

function recordActivity() {
  try {
    fs.mkdirSync(stateDir(), { recursive: true });
    const now = Date.now();
    let stamps = [];
    try {
      stamps = fs
        .readFileSync(activityFile(), 'utf8')
        .split('\n')
        .map(Number)
        .filter((t) => t && now - t < ACTIVITY_WINDOW_MS);
    } catch {
    }
    stamps.push(now);
    fs.writeFileSync(activityFile(), stamps.join('\n'));
  } catch {
  }
}

function activityLevel() {
  try {
    const now = Date.now();
    const count = fs
      .readFileSync(activityFile(), 'utf8')
      .split('\n')
      .map(Number)
      .filter((t) => t && now - t < ACTIVITY_WINDOW_MS).length;
    return Math.min(1, count / ACTIVITY_MAX);
  } catch {
    return 0;
  }
}

function lastActivityMs() {
  try {
    const stamps = fs
      .readFileSync(activityFile(), 'utf8')
      .split('\n')
      .map(Number)
      .filter(Boolean);
    return stamps.length ? Math.max(...stamps) : 0;
  } catch {
    return 0;
  }
}

const WATCHDOG_FRESH_MS = 45000;

function ensureWatchdog() {
  if (process.env.VIBECODE_NO_WATCHDOG) return;
  try {
    if (Date.now() - fs.statSync(watchdogFile()).mtimeMs < WATCHDOG_FRESH_MS) return;
  } catch {
  }
  try {
    const env = { ...process.env };
    delete env.VIBECODE_TOKEN;
    delete env.VIBECODE_EVENT;
    const child = spawn(process.execPath, [path.join(__dirname, 'watchdog.js')], {
      detached: true,
      stdio: 'ignore',
      windowsHide: true,
      env,
    });
    child.on('error', () => {});
    child.unref();
    log('watchdog: spawned');
  } catch {
  }
}

async function play() {
  const token = actionToken();
  await logAudio(`PLAY start token=${token}`);
  if (isDisabled()) {
    log('PLAY skip: disabled flag set');
    return;
  }
  recordIntent(token, 'play');
  const t0 = Date.now();
  recordActivity();
  let wasHalted = true;
  const live = await timed('alive-check', () => alive());
  if (live) {
    wasHalted = (await isHalted()) !== false;
    log(`play: player already alive, resuming (wasHalted=${wasHalted})`);
  } else {
    log('play: no live player, COLD START');
    if (!(await timed('cold-start', () => start(source(), volume())))) {
      log(`play: player did not come up (total ${Date.now() - t0}ms)`);
      return;
    }
  }
  if (superseded(token)) {
    log('play: superseded by a newer event, not resuming');
    return;
  }
  if (wasHalted) {
    const target = targetVolume();
    await send(ipcPath(), { command: ['set_property', 'mute', false] });
    await send(ipcPath(), { command: ['set_property', 'pause', false] });
    const cache = await getProp('demuxer-cache-time');
    if (!(typeof cache === 'number' && cache > 1)) {
      log(`play: cache empty (${cache}), reconnecting fresh`);
      await send(ipcPath(), { command: ['loadfile', source()] });
      await send(ipcPath(), { command: ['set_property', 'mute', false] });
      await send(ipcPath(), { command: ['set_property', 'pause', false] });
    }
    const finished = await timed('fade-in', () => fade(Math.round(target / 2), target, token));
    if (!finished) {
      await send(ipcPath(), { command: ['set_property', 'mute', true] });
      log('play: superseded mid fade-in, re-muted');
    }
  } else if (adaptiveEnabled()) {
    await send(ipcPath(), { command: ['set_property', 'volume', targetVolume()] });
  }
  await logAudio(`PLAY done in ${Date.now() - t0}ms`);
  ensureWatchdog();
}

async function fadeOut() {
  const cur = await getProp('volume');
  if (typeof cur === 'number' && cur > 0) await fade(cur, 0);
}

async function pause() {
  const token = actionToken();
  const t0 = Date.now();
  await logAudio(`PAUSE start token=${token}`);
  if (!focusOn()) {
    log('PAUSE skip: focus mode off (music plays continuously)');
    return;
  }
  recordIntent(token, 'pause');
  const live = await timed('alive-check', () => alive());
  if (!live) {
    log('PAUSE skip: no live player');
    return;
  }
  await send(ipcPath(), { command: ['set_property', 'mute', true] });
  await logAudio(`PAUSE done in ${Date.now() - t0}ms`);
  ensureWatchdog();
}

async function hardPause() {
  if (!(await alive())) return;
  if ((await isHalted()) === false) {
    await fadeOut();
  }
  await send(ipcPath(), { command: ['set_property', 'mute', true] });
  await send(ipcPath(), { command: ['set_property', 'pause', true] });
}

function currentStation() {
  try {
    return fs.readFileSync(stationFile(), 'utf8').trim() || null;
  } catch {
    return null;
  }
}

async function applyStation(url) {
  fs.mkdirSync(stateDir(), { recursive: true });
  fs.writeFileSync(stationFile(), url);
  recordActivity();
  if (await alive()) {
    await send(ipcPath(), { command: ['loadfile', url] });
    await send(ipcPath(), { command: ['set_property', 'volume', 0] });
    await send(ipcPath(), { command: ['set_property', 'mute', false] });
    await send(ipcPath(), { command: ['set_property', 'pause', false] });
    await fadeTo(targetVolume());
  } else {
    await start(url, targetVolume());
    await send(ipcPath(), { command: ['set_property', 'mute', false] });
    await send(ipcPath(), { command: ['set_property', 'pause', false] });
  }
  ensureWatchdog();
}

async function radio(vibe) {
  const url = stations.resolve(vibe);
  if (!url) return;
  await applyStation(url);
}

async function next() {
  const url = stations.nextStation(currentStation());
  if (url) await applyStation(url);
}

async function setVolume(arg) {
  const base = volume();
  let v;
  if (arg === 'up') v = base + 10;
  else if (arg === 'down') v = base - 10;
  else {
    const n = parseInt(arg, 10);
    if (Number.isNaN(n)) return;
    v = n;
  }
  v = Math.max(0, Math.min(100, v));
  fs.mkdirSync(stateDir(), { recursive: true });
  fs.writeFileSync(volumeFile(), String(v));
  if (await alive()) {
    await send(ipcPath(), { command: ['set_property', 'volume', v] });
  }
  log(`volume: ${v}`);
}

async function status() {
  if (isDisabled()) return '';
  const halted = await isHalted();
  if (halted === null) return '';
  return halted ? '❚❚' : '►';
}

async function track() {
  if (isDisabled()) return '';
  const reply = await send(ipcPath(), { command: ['get_property', 'media-title'] });
  if (!reply || reply.error !== 'success' || !reply.data) return '';
  return String(reply.data)
    // eslint-disable-next-line no-control-regex
    .replace(/[\x00-\x1f\x7f\x9b]/g, '')
    .replace(/\s+/g, ' ')
    .trim()
    .slice(0, 48);
}

function stationLabel() {
  try {
    return stations.label(fs.readFileSync(stationFile(), 'utf8').trim());
  } catch {
    return null;
  }
}

function stationTheme() {
  try {
    return stations.theme(fs.readFileSync(stationFile(), 'utf8').trim());
  } catch {
    return null;
  }
}

async function focus(arg) {
  fs.mkdirSync(stateDir(), { recursive: true });
  if (arg === 'off') {
    fs.writeFileSync(noFocusFlag(), '');
    log('focus: off');
    await play();
  } else {
    try {
      fs.unlinkSync(noFocusFlag());
    } catch {
    }
    log('focus: on');
  }
}

function dance() {
  const dancers = [
    ['ヽ(⌐■_■)ノ♪', '♪ヽ(■_■⌐)ノ', 'ヽ(⌐■_■)ノ♬'],
    ['♪┏(・o・)┛', '┗(・o・)┓♪', '♪┏(・o・)┛'],
    ['(♪)┏(＾0＾)┛', '┗(＾0＾)┓(♫)', '(♬)┏(＾0＾)┛'],
    ['⟨♪⟩ ᕕ( ᐛ )ᕗ', 'ᕕ( ᐛ )ᕗ ⟨♫⟩', '⟨♬⟩ ᕕ( ᐛ )ᕗ'],
  ];
  const crew = dancers[Math.floor(Math.random() * dancers.length)];
  const label = stationLabel() || 'vibecode.fm';
  return `\n   ${crew.join('   ')}\n\n   dancing to ${label} — keep coding ♪\n`;
}

function on() {
  log('action=on');
  try {
    fs.unlinkSync(disabledFlag());
  } catch {
  }
}

async function off() {
  log('action=off');
  if (await alive()) {
    await send(ipcPath(), { command: ['quit'] });
  }
  if (process.platform !== 'win32') {
    try {
      fs.unlinkSync(ipcPath());
    } catch {
    }
  }
  fs.mkdirSync(stateDir(), { recursive: true });
  fs.writeFileSync(disabledFlag(), '');
}

module.exports = {
  play,
  pause,
  hardPause,
  radio,
  next,
  setVolume,
  focus,
  dance,
  status,
  track,
  on,
  off,
  activityLevel,
  lastActivityMs,
  adaptiveVolume,
  stationLabel,
  stationTheme,
  stationNames: stations.names,
};

File: src/controller.rs
use crate::ipc;
use crate::log::{log, log_audio};
use crate::paths;
use crate::player;
use crate::stations::{self, Theme};
use serde_json::json;
use std::env;
use std::fs;
use std::process::{Command, Stdio};
use std::thread::sleep;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

const FADE_MS: u64 = 180;
const FADE_STEPS: i64 = 6;

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

fn action_token() -> i64 {
    env::var("VIBECODE_TOKEN")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or_else(now_ms)
}

fn latest_token() -> i64 {
    fs::read_to_string(paths::intent_file())
        .ok()
        .and_then(|s| s.trim().split(' ').next().and_then(|t| t.parse().ok()))
        .unwrap_or(0)
}

fn record_intent(token: i64, action: &str) {
    if token < latest_token() {
        return;
    }
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(paths::intent_file(), format!("{token} {action}"));
}

fn superseded(token: i64) -> bool {
    latest_token() > token
}

fn is_disabled() -> bool {
    paths::disabled_flag().exists()
}

fn focus_on() -> bool {
    !paths::no_focus_flag().exists()
}

fn volume() -> i64 {
    if let Ok(s) = fs::read_to_string(paths::volume_file()) {
        if let Ok(v) = s.trim().parse::<i64>() {
            return v;
        }
    }
    env::var("VIBECODE_VOLUME")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(70)
}

fn source() -> String {
    if let Ok(s) = env::var("VIBECODE_SOURCE") {
        return s;
    }
    if let Ok(s) = fs::read_to_string(paths::station_file()) {
        let t = s.trim();
        if !t.is_empty() {
            return t.to_string();
        }
    }
    paths::default_source()
}

fn is_halted() -> Option<bool> {
    let pause = ipc::get_prop("pause")?;
    if pause.as_bool() == Some(true) {
        return Some(true);
    }
    Some(ipc::get_prop("mute").and_then(|m| m.as_bool()) == Some(true))
}

fn fade(from: i64, to: i64, guard: Option<i64>) -> bool {
    for i in 1..=FADE_STEPS {
        if let Some(g) = guard {
            if superseded(g) {
                return false;
            }
        }
        let v = (from as f64 + (to - from) as f64 * i as f64 / FADE_STEPS as f64).round() as i64;
        ipc::set_prop("volume", json!(v));
        sleep(Duration::from_millis(FADE_MS / FADE_STEPS as u64));
    }
    true
}

fn fade_to(target: i64) {
    fade(0, target, None);
}

fn record_activity() {
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(paths::activity_file(), now_ms().to_string());
}

pub fn last_activity_ms() -> i64 {
    fs::read_to_string(paths::activity_file())
        .ok()
        .and_then(|s| s.trim().parse().ok())
        .unwrap_or(0)
}

fn ensure_watchdog() {
    if env::var("VIBECODE_NO_WATCHDOG").is_ok() {
        return;
    }
    if let Ok(meta) = fs::metadata(paths::watchdog_file()) {
        if let Ok(modified) = meta.modified() {
            if SystemTime::now()
                .duration_since(modified)
                .is_ok_and(|d| d.as_secs() < 45)
            {
                return;
            }
        }
    }
    let exe = match env::current_exe() {
        Ok(e) => e,
        Err(_) => return,
    };
    let mut cmd = Command::new(exe);
    cmd.arg("watchdog");
    cmd.env_remove("VIBECODE_TOKEN")
        .env_remove("VIBECODE_EVENT");
    cmd.stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    let _ = player::spawn_detached(&mut cmd);
    log("watchdog: spawned");
}

pub fn play() {
    let token = action_token();
    log_audio(&format!("PLAY start token={token}"));
    if is_disabled() {
        log("PLAY skip: disabled flag set");
        return;
    }
    record_intent(token, "play");
    let t0 = now_ms();
    record_activity();
    let mut was_halted = true;
    if player::alive() {
        was_halted = is_halted() != Some(false);
    } else if !player::start(&source(), volume()) {
        return;
    }
    if superseded(token) {
        log("play: superseded by a newer event, not resuming");
        return;
    }
    if was_halted {
        let target = volume();
        ipc::set_prop("mute", json!(false));
        ipc::set_prop("pause", json!(false));
        let warm = ipc::get_prop("demuxer-cache-time")
            .and_then(|v| v.as_f64())
            .map(|c| c > 1.0)
            .unwrap_or(false);
        if !warm {
            ipc::command(json!(["loadfile", source()]));
            ipc::set_prop("mute", json!(false));
            ipc::set_prop("pause", json!(false));
        }
        if !fade((target as f64 / 2.0).round() as i64, target, Some(token)) {
            ipc::set_prop("mute", json!(true));
        }
    }
    log_audio(&format!("PLAY done in {}ms", now_ms() - t0));
    ensure_watchdog();
}

pub fn pause() {
    let token = action_token();
    let t0 = now_ms();
    log_audio(&format!("PAUSE start token={token}"));
    if !focus_on() {
        log("PAUSE skip: focus mode off");
        return;
    }
    record_intent(token, "pause");
    if !player::alive() {
        log("PAUSE skip: no live player");
        return;
    }
    ipc::set_prop("mute", json!(true));
    log_audio(&format!("PAUSE done in {}ms", now_ms() - t0));
    ensure_watchdog();
}

pub fn quit_player() {
    if player::alive() {
        ipc::command(json!(["quit"]));
    }
    #[cfg(unix)]
    let _ = fs::remove_file(paths::ipc_path());
}

fn current_station() -> Option<String> {
    fs::read_to_string(paths::station_file())
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
}

fn apply_station(url: &str) {
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(paths::station_file(), url);
    record_activity();
    if player::alive() {
        ipc::command(json!(["loadfile", url]));
        ipc::set_prop("volume", json!(0));
        ipc::set_prop("mute", json!(false));
        ipc::set_prop("pause", json!(false));
        fade_to(volume());
    } else {
        player::start(url, volume());
        ipc::set_prop("mute", json!(false));
        ipc::set_prop("pause", json!(false));
    }
    ensure_watchdog();
}

pub fn radio(vibe: Option<&str>) {
    if let Some(url) = vibe.and_then(stations::resolve) {
        apply_station(&url);
    }
}

pub fn next() {
    if let Some(url) = stations::next_station(current_station().as_deref()) {
        apply_station(&url);
    }
}

pub fn set_volume(arg: Option<&str>) {
    let base = volume();
    let v = match arg {
        Some("up") => base + 10,
        Some("down") => base - 10,
        Some(n) => match n.parse::<i64>() {
            Ok(x) => x,
            Err(_) => return,
        },
        None => return,
    }
    .clamp(0, 100);
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(paths::volume_file(), v.to_string());
    if player::alive() {
        ipc::set_prop("volume", json!(v));
    }
    log(&format!("volume: {v}"));
}

pub fn focus(arg: Option<&str>) {
    let _ = fs::create_dir_all(paths::state_dir());
    if arg == Some("off") {
        let _ = fs::write(paths::no_focus_flag(), "");
        log("focus: off");
        play();
    } else {
        let _ = fs::remove_file(paths::no_focus_flag());
        log("focus: on");
    }
}

pub fn minimal(arg: Option<&str>) {
    let _ = fs::create_dir_all(paths::state_dir());
    let on = match arg {
        Some("on") => true,
        Some("off") => false,
        _ => !paths::minimal_flag().exists(),
    };
    if on {
        let _ = fs::write(paths::minimal_flag(), "");
        log("minimal: on");
    } else {
        let _ = fs::remove_file(paths::minimal_flag());
        log("minimal: off");
    }
}

pub fn minimal_active() -> bool {
    paths::minimal_flag().exists()
}

pub fn status() -> String {
    if is_disabled() {
        return String::new();
    }
    match is_halted() {
        None => String::new(),
        Some(true) => "❚❚".into(),
        Some(false) => "►".into(),
    }
}

fn sanitize_title(data: &str) -> String {
    let cleaned: String = data
        .chars()
        .filter(|c| !c.is_control() && *c != '\u{9b}')
        .collect();
    cleaned
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
        .chars()
        .take(48)
        .collect()
}

pub fn track() -> String {
    if is_disabled() {
        return String::new();
    }
    match ipc::get_prop("media-title").and_then(|v| v.as_str().map(String::from)) {
        Some(s) if !s.is_empty() => sanitize_title(&s),
        _ => String::new(),
    }
}

pub fn station_label() -> Option<String> {
    stations::label(&current_station()?)
}

pub fn station_theme() -> Option<Theme> {
    stations::theme(&current_station()?)
}

pub fn on() {
    log("action=on");
    let _ = fs::remove_file(paths::disabled_flag());
}

pub fn off() {
    log("action=off");
    quit_player();
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(paths::disabled_flag(), "");
}

pub fn debug(arg: Option<&str>) {
    let _ = fs::create_dir_all(paths::state_dir());
    if arg == Some("off") {
        let _ = fs::remove_file(paths::debug_flag());
    } else {
        let _ = fs::write(paths::debug_flag(), "");
    }
}

pub fn setup_statusline(arg: Option<&str>) {
    let path = paths::settings_file();
    let mut root: serde_json::Value = match fs::read_to_string(&path) {
        Ok(s) => match serde_json::from_str(&s) {
            Ok(v) => v,
            Err(_) => {
                eprintln!("settings.json is not valid JSON; leaving it untouched");
                return;
            }
        },
        Err(_) => serde_json::json!({}),
    };
    let Some(obj) = root.as_object_mut() else {
        return;
    };
    if arg == Some("off") {
        obj.remove("statusLine");
    } else {
        let exe = env::current_exe()
            .map(|p| p.to_string_lossy().into_owned())
            .unwrap_or_else(|_| "vibecode-fm".into());
        obj.insert(
            "statusLine".into(),
            json!({ "type": "command", "command": format!("\"{exe}\" statusline") }),
        );
    }
    if let Ok(pretty) = serde_json::to_string_pretty(&root) {
        if let Some(dir) = path.parent() {
            let _ = fs::create_dir_all(dir);
        }
        let tmp = path.with_extension("json.tmp");
        if fs::write(&tmp, pretty).is_ok() {
            let _ = fs::rename(&tmp, &path);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_strips_control_and_escape_chars() {
        let out = sanitize_title("Evil\x1b[31m\x07\x00 Song");
        assert!(!out.chars().any(|c| c.is_control()));
        assert_eq!(out, "Evil[31m Song");
    }
}
